FileHistory mode at browse startup

gerhardol opened the pull request:
Final (?) part of #9796 

## Proposed changes

When starting up with command line arguments "filehistory" or
"blamehistory" (for instance from VS or explorer),
hide the side panel and start in FileTree tab to
provide an experience similar to old FileHistory.

The forced side panel state is not persisted to settings at shutdown.

If argument is "blamehistory" the Blame viewer is explicitly enabled
(similar to blame from DifTab), for "filehistory" the last used form
is used.

## Screenshots <!-- Remove this section if PR does not change UI -->

When starting with command line arguments "filehistory appveyor.yml"

### Before

![image](https://user-images.githubusercontent.com/6248932/153090698-2d28eb3d-ab1a-4e80-8b20-d1054849491a.png)

### After

![image](https://user-images.githubusercontent.com/6248932/153090665-27f2daef-d0dd-4136-be1d-6087b7fa9e61.png)

## Test methodology <!-- How did you ensure quality? -->

manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mstv reviewed (commented):
Almost there with 93c62b710ab2b484b538a92a94f26515b391c8b9

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.BrowseArguments.cs:
```diff
@@ -25,6 +25,11 @@ public record BrowseArguments
+            /// <summary>
+            /// If to start in FileHistory mode, including starting as default or requesting blame view
+            /// </summary>
+            public bool? IsFileBlameHistory { get; init; }
```
I avoid that to avoid putting the enum in a separate file.

gerhardol replied:
I will remove the file/blame difference and use a plain bool.

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.InitRevisionGrid.cs:
```diff
@@ -40,7 +40,12 @@ private void InitRevisionGrid(ObjectId? selectedId, ObjectId? firstId)
+                {
+                    // Start in "filehistory" mode
+                    CommitInfoTabControl.SelectedTab = TreeTabPage;
+                }
```
1. Only handle this once by setting isBlame to null (the value is captured)
2. At startup HEAD is selected or possibly an ordinary commit the user has selected, so never artificial
(I considered adding a check but that would need to be in sync with the hiding of the tab anyway so no point).

> Almost there with https://github.com/gitextensions/gitextensions/commit/93c62b710ab2b484b538a92a94f26515b391c8b9

Thanks, I had skipped that thinking that this was only needed from the (old) command line arguments...

mstv replied:
(I use the standalone history instance f.i. when diffing two branches.)

gerhardol replied:
> (I use the standalone history instance f.i. when diffing two branches.)

I have stopped doing that and try to keep number of GE instances below 5, used to be about 10 at a time.
But of course that will be supported still.

mstv replied:
> At startup HEAD is selected or possibly an ordinary commit the user has selected, so never artificial

This is not true: Hit `H` on an uncommitted file.
We continue the discussion which tab to open, after this has been merged.

gerhardol replied:
Correct, now just Diff is shown. Adding to discussions

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -253,6 +255,13 @@ public FormBrowse(GitUICommands commands, BrowseArguments args)
             SystemEvents.SessionEnding += (sender, args) => SaveApplicationSettings();
 
+            _isFileBlameHistory = args.IsFileBlameHistory;
+            if (_isFileBlameHistory is not null && _isFileBlameHistory.Value)
```
If the `bool?` is kept instead of an `enum`:
```suggestion
            if (_isFileBlameHistory is true)
```

gerhardol replied:
ok

gerhardol commented:
Plan to squash and merge this in two days

RussKie commented:
Do we have a mechanism to pass a specific hash to blame?

Also I think we should not show the file tree for the blame or history
modes. That is, if I blame a file I'm generally interested in the history
for this file. The file tree is a distraction just like the left panel
(kudos for hiding it).

On Sat, 12 Feb 2022, 8:16 am Gerhard Olsson, ***@***.***>
wrote:

> Plan to squash and merge this in two days
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/9862#issuecomment-1036630073>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXUAISU4M54SZSUO2K3U2V4DXANCNFSM5N33LPQQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

RussKie commented:
I also like the suggestion with the enum. I think it makes it clearer.

On Sun, 13 Feb 2022, 9:49 am Igor Velikorossov, ***@***.***> wrote:

> Do we have a mechanism to pass a specific hash to blame?
>
> Also I think we should not show the file tree for the blame or history
> modes. That is, if I blame a file I'm generally interested in the history
> for this file. The file tree is a distraction just like the left panel
> (kudos for hiding it).
>
> On Sat, 12 Feb 2022, 8:16 am Gerhard Olsson, ***@***.***>
> wrote:
>
>> Plan to squash and merge this in two days
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/gitextensions/gitextensions/pull/9862#issuecomment-1036630073>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABBTEXUAISU4M54SZSUO2K3U2V4DXANCNFSM5N33LPQQ>
>> .
>> You are receiving this because you are subscribed to this thread.Message
>> ID: ***@***.***>
>>
>

gerhardol commented:
Can we do further changes in a follow up PR instead?

> > Also I think we should not show the file tree for the blame or history modes. That is, if I blame a file I'm generally interested in the history for this file. The file tree is a distraction just like the left panel (kudos for hiding it).

Then you loose the actins for the file (switch blame/view etc - hotkeys are available though).
I also want to see blame for other files.

The filetree could be collapsed or since there is no simple way to to this now, be minimized in size (with hacks to not save size similar to for sidepanel).

> Valid point. I use to switch to the Diff tab, too. This also works around that the FileTree is not available for artificial commits.

That would require a new tab that is shown instead, FileTree is only for real commits.
The blame-in-artificial is for real commits, it omits the local changes.

gerhardol commented:
Plan to merge as is tomorrow

gerhardol commented:
Squashed, updated commit message and a few comments, changed `_fileBlameHistorySidePanelStartupState` to `bool`
Applied a similar change as in https://github.com/gitextensions/gitextensions/pull/9862#pullrequestreview-878052209 when starting without `-commit`

Remaining discussions, to be tracked in #9796 (already mentioned there).
* https://github.com/gitextensions/gitextensions/pull/9862#discussion_r805425757 How to handle file history for artificial commits. Disable action, show blame in difftab or switch to HEAD.
* https://github.com/gitextensions/gitextensions/pull/9862#issuecomment-1037547666 Hide the fileviewer for filetree (and diffviewer)
